README.md
# 🏙 Rocket Dog

A dark and light Neovim theme written in Lua ported from the Visual Studio Code
[tokyonight](https://github.com/folke/tokyonight.nvim) theme.

## ✨ Features

- supports the latest Neovim 5.0 features like TreeSitter and LSP
- vim terminal colors
- darker background for sidebar-like windows

### Plugin Support

- [TreeSitter](https://github.com/nvim-treesitter/nvim-treesitter)
- [LSP Diagnostics](https://neovim.io/doc/user/lsp.html)
- [LSP Trouble](https://github.com/folke/lsp-trouble.nvim)
- [LSP Saga](https://github.com/glepnir/lspsaga.nvim)
- [Git Signs](https://github.com/lewis6991/gitsigns.nvim)
- [Git Gutter](https://github.com/airblade/vim-gitgutter)
- [Telescope](https://github.com/nvim-telescope/telescope.nvim)
- [NvimTree](https://github.com/kyazdani42/nvim-tree.lua)
- [WhichKey](https://github.com/liuchengxu/vim-which-key)
- [Indent Blankline](https://github.com/lukas-reineke/indent-blankline.nvim)
- [Dashboard](https://github.com/glepnir/dashboard-nvim)
- [BufferLine](https://github.com/akinsho/nvim-bufferline.lua)
- [Lualine](https://github.com/hoob3rt/lualine.nvim)
- [Neogit](https://github.com/TimUntersberger/neogit)
- [vim-sneak](https://github.com/justinmk/vim-sneak)
- [Fern](https://github.com/lambdalisue/fern.vim)
- [Barbar](https://github.com/romgrk/barbar.nvim)
- [Scrollbar](https://github.com/petertriho/nvim-scrollbar)
- [Mini](https://github.com/echasnovski/mini.nvim)

## ⚡️ Requirements

- Neovim >= 0.6.0

## 📦 Installation

Install the theme with your preferred package manager:

[vim-plug](https://github.com/junegunn/vim-plug)

```vim
Plug 'lessthanseventy/rocket_dog.nvim', { 'branch': 'main' }
```

[packer](https://github.com/wbthomason/packer.nvim)

```lua
use 'lessthanseventy/rocket_dog.nvim'
```

[lazy](https://github.com/folke/lazy.nvim)

```lua
  {
    "lessthanseventy/rocket_dog.nvim",
    config = function()
      require("rocket_dog").setup({
        style = "night",        -- The theme comes in three styles, `storm`, `moon`, a darker variant `night` and `day`
        light_style = "day",    -- The theme is used when the background is set to light
        terminal_colors = true, -- Configure the colors used when opening a `:terminal` in Neovim
        styles = {
          -- Style to be applied to different syntax groups
          -- Value is any valid attr-list value for `:help nvim_set_hl`
          comments = { italic = false },
          keywords = { italic = false },
          functions = {},
          variables = {},
          -- Background styles. Can be "dark", "transparent" or "normal"
          sidebars = "dark",         -- style for sidebars, see below
          floats = "dark",           -- style for floating windows
        },
        sidebars = { "qf", "help" }, -- Set a darker background on sidebar-like windows. For example: `["qf", "vista_kind", "terminal", "packer"]`
      })
      vim.cmd("colorscheme rocket_dog")
    end,
  },
```

## 🚀 Usage

Enable the colorscheme:

```vim
" Vim Script
colorscheme rocket_dog

" There are also colorschemes for the different styles
colorscheme rocket_dog-night
colorscheme rocket_dog-storm
colorscheme rocket_dog-day
colorscheme rocket_dog-moon
```

```lua
-- Lua
vim.cmd[[colorscheme rocket_dog]]
```

To enable the `rocket_dog` theme for `Barbecue`:

```lua
require('barbecue').setup {
  -- ... your barbecue config
  theme = 'rocket_dog',
  -- ... your barbecue config
}
```

To enable the `rocket_dog` theme for `Lualine`, simply specify it in your
lualine settings:

```lua
require('lualine').setup {
  options = {
    -- ... your lualine config
    theme = 'rocket_dog'
    -- ... your lualine config
  }
}
```

To enable the `rocket_dog` colorscheme for `Lightline`:

```vim
" Vim Script
let g:lightline = {'colorscheme': 'rocket_dog'}
```

## ⚙️ Configuration

> ❗️ configuration needs to be set **BEFORE** loading the color scheme with
> `colorscheme rocket_dog`

The theme comes in four styles, `storm`, `moon`, a darker variant `night` and `day`.

The **day** style will be used if:

- `{ style = "day"}` was passed to `setup(options)`
- or `vim.o.background = "light"`

rocket_dog will use the default options, unless you call `setup`.

```lua
require("rocket_dog").setup({
  -- your configuration comes here
  -- or leave it empty to use the default settings
  style = "storm", -- The theme comes in three styles, `storm`, `moon`, a darker variant `night` and `day`
  light_style = "day", -- The theme is used when the background is set to light
  transparent = false, -- Enable this to disable setting the background color
  terminal_colors = true, -- Configure the colors used when opening a `:terminal` in Neovim
  styles = {
    -- Style to be applied to different syntax groups
    -- Value is any valid attr-list value for `:help nvim_set_hl`
    comments = { italic = true },
    keywords = { italic = true },
    functions = {},
    variables = {},
    -- Background styles. Can be "dark", "transparent" or "normal"
    sidebars = "dark", -- style for sidebars, see below
    floats = "dark", -- style for floating windows
  },
  sidebars = { "qf", "help" }, -- Set a darker background on sidebar-like windows. For example: `["qf", "vista_kind", "terminal", "packer"]`
  day_brightness = 0.3, -- Adjusts the brightness of the colors of the **Day** style. Number between 0 and 1, from dull to vibrant colors
  hide_inactive_statusline = false, -- Enabling this option, will hide inactive statuslines and replace them with a thin border instead. Should work with the standard **StatusLine** and **LuaLine**.
  dim_inactive = false, -- dims inactive windows
  lualine_bold = false, -- When `true`, section headers in the lualine theme will be bold

  --- You can override specific color groups to use other groups or a hex color
  --- function will be called with a ColorScheme table
  ---@param colors ColorScheme
  on_colors = function(colors) end,

  --- You can override specific highlights to use other groups or a hex color
  --- function will be called with a Highlights and ColorScheme table
  ---@param highlights Highlights
  ---@param colors ColorScheme
  on_highlights = function(highlights, colors) end,
})
```

## 🪓 Overriding Colors & Highlight Groups

How the highlight groups are calculated:

1. the **colors** for the style are calculated based on your config
2. `config.on_colors(colors)` is ran, where you can override the colors
3. the **colors** are then used to generate the highlight groups
4. `config.on_highlights(highlights, colors)` is ran, where you can overide the highlight groups

Please refer to default values for `colors` and `highlights` for the [storm](extras/lua/rocket_dog_storm.lua), [moon](extras/lua/rocket_dog_moon.lua), [night](extras/lua/rocket_dog_night.lua), [day](extras/lua/rocket_dog_day.lua)

Example for changing some settings and colors

```lua
require("rocket_dog").setup({
  -- use the night style
  style = "night",
  -- disable italic for functions
  styles = {
    functions = {}
  },
  sidebars = { "qf", "vista_kind", "terminal", "packer" },
  -- Change the "hint" color to the "orange" color, and make the "error" color bright red
  on_colors = function(colors)
    colors.hint = colors.orange
    colors.error = "#ff0000"
  end
})
```

Example to make Telescope
[borderless](https://github.com/nvim-telescope/telescope.nvim/wiki/Gallery#borderless)

```lua
require("rocket_dog").setup({
  on_highlights = function(hl, c)
    local prompt = "#2d3149"
    hl.TelescopeNormal = {
      bg = c.bg_dark,
      fg = c.fg_dark,
    }
    hl.TelescopeBorder = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
    hl.TelescopePromptNormal = {
      bg = prompt,
    }
    hl.TelescopePromptBorder = {
      bg = prompt,
      fg = prompt,
    }
    hl.TelescopePromptTitle = {
      bg = prompt,
      fg = prompt,
    }
    hl.TelescopePreviewTitle = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
    hl.TelescopeResultsTitle = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
  end,
})
```

### Making `undercurls` work properly in **Tmux**

To have undercurls show up and in color, add the following to your **Tmux**
config file:

```sh
# Undercurl
set -g default-terminal "${TERM}"
set -as terminal-overrides ',*:Smulx=\E[4::%p1%dm'  # undercurl support
set -as terminal-overrides ',*:Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m'  # underscore colours - needs tmux-3.0
```

## 🍭 Extras

Extra color configs for **Kitty**, **Alacritty**, **Fish**, **WezTerm**,
**iTerm** and **foot** can be found in [extras](extras/). To use them, refer to
their respective documentation.

![image](https://user-images.githubusercontent.com/292349/115395546-d8d6f880-a198-11eb-98fb-a1194787701d.png)

You can easily use the color palette for other plugins inside your Neovim
config:

```lua
local colors = require("rocket_dog.colors").setup() -- pass in any of the config options as explained above
local util = require("rocket_dog.util")

aplugin.background = colors.bg_dark
aplugin.my_error = util.brighten(colors.red1, 0.3)
```

## 🔥 Contributing

Pull requests are welcome. For the `extras`, we use a simple template system
that can be used to generate themes for the different styles.

How to add a new extra template:

1. create a file like `lua/rocket_dog/extra/cool-app.lua`
2. add the name and output file extension to the `extras` table in
   `lua/rocket_dog/extra/init.lua`
3. in Nvim, run `:lua require("rocket_dog.extra").setup()` to generate / update
   extra themes
4. commit the newly created themes under `extra/`


CHANGELOG.md
# Changelog

## [1.3.0](https://github.com/lessthanseventy/tokyonight.nvim/compare/v1.2.0...v1.3.0) (2023-02-09)


### Features

* **extras:** add fish theme format ([#306](https://github.com/lessthanseventy/tokyonight.nvim/issues/306)) ([3685c24](https://github.com/lessthanseventy/tokyonight.nvim/commit/3685c241a2adaad70299989d32f67fc5c482ff3c))


### Bug Fixes

* treesitter todo keyword highlight ([#310](https://github.com/lessthanseventy/tokyonight.nvim/issues/310)) ([0c21ef1](https://github.com/lessthanseventy/tokyonight.nvim/commit/0c21ef11b8bbe2c6ec68a69553bee75a75e1587b))

## [1.2.0](https://github.com/lessthanseventy/tokyonight.nvim/compare/v1.1.0...v1.2.0) (2023-02-07)


### Features

* **wezterm:** added tab_bar colors. only works with the retro tab bar. Fixes [#304](https://github.com/lessthanseventy/tokyonight.nvim/issues/304) ([56945bd](https://github.com/lessthanseventy/tokyonight.nvim/commit/56945bd0d312dc3ed84466d7a6cbfc5e44fbbb4e))

## [1.1.0](https://github.com/lessthanseventy/tokyonight.nvim/compare/v1.0.0...v1.1.0) (2023-01-23)


### Features

* **extras:** added prism.js themes ([14ca396](https://github.com/lessthanseventy/tokyonight.nvim/commit/14ca396af092674ec35b04a8da097e2b223d4de6))
* **prism:** better colors ([7c22129](https://github.com/lessthanseventy/tokyonight.nvim/commit/7c22129e3bda1aa08734a1a94a8a69193a538378))


### Bug Fixes

* **lazy:** set todo progress to fg_gutter ([655acfa](https://github.com/lessthanseventy/tokyonight.nvim/commit/655acfad615af196607e29fcabd8998ade9489a5))
* **sublime:** don't hardcode background and foreground colours in Sublime theme ([#298](https://github.com/lessthanseventy/tokyonight.nvim/issues/298)) ([10730e6](https://github.com/lessthanseventy/tokyonight.nvim/commit/10730e6903b1399aede2cad0b93cee4b84b25cda))

## 1.0.0 (2023-01-04)


### ⚠ BREAKING CHANGES

* move extras in seperate directories
* new `moon` style for TokyoNight, based on the Moonlight theme
* added docs for the new settings

### Features

* "day" variant ([8889ad8](https://github.com/lessthanseventy/tokyonight.nvim/commit/8889ad88481cd07b9f32eab2879ab6a32a6cc75a))
* add bg hightlight on NvimTreeOpenedFile ([#163](https://github.com/lessthanseventy/tokyonight.nvim/issues/163)) ([a1d1c0f](https://github.com/lessthanseventy/tokyonight.nvim/commit/a1d1c0fb19702f597cb16ef628baddff64ac2585))
* add color config for foot ([#147](https://github.com/lessthanseventy/tokyonight.nvim/issues/147)) ([eaf6a76](https://github.com/lessthanseventy/tokyonight.nvim/commit/eaf6a7607e2ce949fe53981ddbde92dd45af3d06))
* add color for Nvimtree vertsplit ([#215](https://github.com/lessthanseventy/tokyonight.nvim/issues/215)) ([2a74abd](https://github.com/lessthanseventy/tokyonight.nvim/commit/2a74abdf1206a015a8782e6cf33a94142b5cbdd0))
* Add colors to Neotest ([#229](https://github.com/lessthanseventy/tokyonight.nvim/issues/229)) ([66bfc2e](https://github.com/lessthanseventy/tokyonight.nvim/commit/66bfc2e8f754869c7b651f3f47a2ee56ae557764))
* add colorscheme generator and themes for xfce4 terminal ([#81](https://github.com/lessthanseventy/tokyonight.nvim/issues/81)) ([0b4fdc5](https://github.com/lessthanseventy/tokyonight.nvim/commit/0b4fdc5d63be8c9c005e125ecbd55012bbf41174))
* Add explicit support for 'mini.nvim'. ([#179](https://github.com/lessthanseventy/tokyonight.nvim/issues/179)) ([8583ca3](https://github.com/lessthanseventy/tokyonight.nvim/commit/8583ca39dd4098e09cbb66178cc213cb864a0063))
* add hl `LspInfoBorder` ([#236](https://github.com/lessthanseventy/tokyonight.nvim/issues/236)) ([a950f62](https://github.com/lessthanseventy/tokyonight.nvim/commit/a950f6215772d3ee8db61720b7bda43551e56c40))
* add lightline tab colors ([#69](https://github.com/lessthanseventy/tokyonight.nvim/issues/69)) ([4d45b01](https://github.com/lessthanseventy/tokyonight.nvim/commit/4d45b01c46c26350586a92319b62bf35a1ee7765))
* add lualine terminal mode ([#263](https://github.com/lessthanseventy/tokyonight.nvim/issues/263)) ([a7e0108](https://github.com/lessthanseventy/tokyonight.nvim/commit/a7e010826ef8ab90f9b37396954a1533c9178732))
* add new barbar's highlight groups ([#291](https://github.com/lessthanseventy/tokyonight.nvim/issues/291)) ([03c03ed](https://github.com/lessthanseventy/tokyonight.nvim/commit/03c03eddace01bfe127f0a6d9413f84a960ea435))
* add options light_style ([#231](https://github.com/lessthanseventy/tokyonight.nvim/issues/231)) ([df281cc](https://github.com/lessthanseventy/tokyonight.nvim/commit/df281cc0f1a277b6150df24feb809e86f57c2374))
* add setting for bold A section of lualine ([#68](https://github.com/lessthanseventy/tokyonight.nvim/issues/68)) ([933bd15](https://github.com/lessthanseventy/tokyonight.nvim/commit/933bd158ce48d853548cd290e246636fa4d55fd4))
* add support for `vim.diagnostic` and `nvim-cmp` custom menu ([#103](https://github.com/lessthanseventy/tokyonight.nvim/issues/103)) ([cd6f57c](https://github.com/lessthanseventy/tokyonight.nvim/commit/cd6f57c0a419979686bc507b77c9c281e57e4d46))
* add support for nvim-scrollbar ([#141](https://github.com/lessthanseventy/tokyonight.nvim/issues/141)) ([0786a2a](https://github.com/lessthanseventy/tokyonight.nvim/commit/0786a2acd80bee5d0c61b53da77a0a6305b333ab))
* add support for ts-rainbow ([#174](https://github.com/lessthanseventy/tokyonight.nvim/issues/174)) ([ef3abe6](https://github.com/lessthanseventy/tokyonight.nvim/commit/ef3abe666869f43683ec8b5cb9eab9089462ffcc))
* add template for iterm2 ([#199](https://github.com/lessthanseventy/tokyonight.nvim/issues/199)) ([54cf70d](https://github.com/lessthanseventy/tokyonight.nvim/commit/54cf70dbd19acd7c32cc12bb743fec17eaf6fef5))
* Add themes for terminator terminal ([#264](https://github.com/lessthanseventy/tokyonight.nvim/issues/264)) ([a6d34b4](https://github.com/lessthanseventy/tokyonight.nvim/commit/a6d34b4e6df02e5b90a787d16d6d445d39546acb))

lua/rocket_dog/util.lua
local ts = require("rocket_dog.treesitter")

local M = {}

M.bg = "#000000"
M.fg = "#ffffff"
M.day_brightness = 0.3

---@param c  string
local function hexToRgb(c)
  c = string.lower(c)
  return { tonumber(c:sub(2, 3), 16), tonumber(c:sub(4, 5), 16), tonumber(c:sub(6, 7), 16) }
end

---@param foreground string foreground color
---@param background string background color
---@param alpha number|string number between 0 and 1. 0 results in bg, 1 results in fg
function M.blend(foreground, background, alpha)
  alpha = type(alpha) == "string" and (tonumber(alpha, 16) / 0xff) or alpha
  local bg = hexToRgb(background)
  local fg = hexToRgb(foreground)

  local blendChannel = function(i)
    local ret = (alpha * fg[i] + ((1 - alpha) * bg[i]))
    return math.floor(math.min(math.max(0, ret), 255) + 0.5)
  end

  return string.format("#%02x%02x%02x", blendChannel(1), blendChannel(2), blendChannel(3))
end

function M.darken(hex, amount, bg)
  return M.blend(hex, bg or M.bg, amount)
end

function M.lighten(hex, amount, fg)
  return M.blend(hex, fg or M.fg, amount)
end

function M.invert_color(color)
  local hsluv = require("rocket_dog.hsluv")
  if color ~= "NONE" then
    local hsl = hsluv.hex_to_hsluv(color)
    hsl[3] = 100 - hsl[3]
    if hsl[3] < 40 then
      hsl[3] = hsl[3] + (100 - hsl[3]) * M.day_brightness
    end
    return hsluv.hsluv_to_hex(hsl)
  end
  return color
end

---@param group string
function M.highlight(group, hl)
  group = ts.get(group)
  if not group then
    return
  end
  if hl.style then
    if type(hl.style) == "table" then
      hl = vim.tbl_extend("force", hl, hl.style)
    elseif hl.style:lower() ~= "none" then
      -- handle old string style definitions
      for s in string.gmatch(hl.style, "([^,]+)") do
        hl[s] = true
      end
    end
    hl.style = nil
  end
  vim.api.nvim_set_hl(0, group, hl)
end

--- Delete the autocmds when the theme changes to something else
function M.onColorScheme()
  vim.cmd([[autocmd! rocket_dog]])
  vim.cmd([[augroup! rocket_dog]])
end

---@param config Config
function M.autocmds(config)
  vim.cmd([[augroup rocket_dog]])
  vim.cmd([[  autocmd!]])
  vim.cmd([[  autocmd ColorSchemePre * lua require("rocket_dog.util").onColorScheme()]])

  vim.cmd(
    [[  autocmd FileType ]]
      .. table.concat(config.sidebars, ",")
      .. [[ setlocal winhighlight=Normal:NormalSB,SignColumn:SignColumnSB]]
  )
  if vim.tbl_contains(config.sidebars, "terminal") then
    vim.cmd([[  autocmd TermOpen * setlocal winhighlight=Normal:NormalSB,SignColumn:SignColumnSB]])
  end
  vim.cmd([[augroup end]])
end

-- Simple string interpolation.
--
-- Example template: "${name} is ${value}"
--
---@param str string template string
---@param table table key value pairs to replace in the string
function M.template(str, table)
  return (
    str:gsub("($%b{})", function(w)
      return vim.tbl_get(table, unpack(vim.split(w:sub(3, -2), ".", { plain = true }))) or w
    end)
  )
end

function M.syntax(syntax)
  for group, colors in pairs(syntax) do
    M.highlight(group, colors)
  end
end

---@param colors ColorScheme
function M.terminal(colors)
  -- dark
  vim.g.terminal_color_0 = colors.black
  vim.g.terminal_color_8 = colors.terminal_black

  -- light
  vim.g.terminal_color_7 = colors.fg_dark
  vim.g.terminal_color_15 = colors.fg

  -- colors
  vim.g.terminal_color_1 = colors.red
  vim.g.terminal_color_9 = colors.red

  vim.g.terminal_color_2 = colors.green
  vim.g.terminal_color_10 = colors.green

  vim.g.terminal_color_3 = colors.yellow
  vim.g.terminal_color_11 = colors.yellow

  vim.g.terminal_color_4 = colors.blue
  vim.g.terminal_color_12 = colors.blue

  vim.g.terminal_color_5 = colors.magenta
  vim.g.terminal_color_13 = colors.magenta

  vim.g.terminal_color_6 = colors.cyan
  vim.g.terminal_color_14 = colors.cyan
end

---@param colors ColorScheme
function M.invert_colors(colors)
  if type(colors) == "string" then
    ---@diagnostic disable-next-line: return-type-mismatch
    return M.invert_color(colors)
  end
  for key, value in pairs(colors) do
    colors[key] = M.invert_colors(value)
  end
  return colors
end

---@param hls Highlights
function M.invert_highlights(hls)
  for _, hl in pairs(hls) do
    if hl.fg then
      hl.fg = M.invert_color(hl.fg)
    end
    if hl.bg then
      hl.bg = M.invert_color(hl.bg)
    end
    if hl.sp then
      hl.sp = M.invert_color(hl.sp)
    end
  end
end

---@param theme Theme
function M.load(theme)
  -- only needed to clear when not the default colorscheme
  if vim.g.colors_name then
    vim.cmd("hi clear")
  end

  vim.o.termguicolors = true
  vim.g.colors_name = "rocket_dog"

  if ts.new_style() then
    for group, colors in pairs(ts.defaults) do
      if not theme.highlights[group] then
        M.highlight(group, colors)
      end
    end
  end

  M.syntax(theme.highlights)

  -- vim.api.nvim_set_hl_ns(M.ns)
  if theme.config.terminal_colors then
    M.terminal(theme.colors)
  end

  M.autocmds(theme.config)

  vim.defer_fn(function()
    M.syntax(theme.defer)
  end, 100)
end

return M
